### ISSUE TITLE
BatchNormalization and SpatialBatchNormalization fail assert for test phase

### ISSUE THREAD

**Issue by hnanhtuan** [torch/nn#903]
Hi everyone,

I think there is a bug in "BatchNormalization.lua". 

```
qlua: ...an/torch/install/share/lua/5.1/nn/BatchNormalization.lua:77: only mini-batch supported (4D tensor), got 3D tensor instead
stack traceback:
	[C]: at 0x7f120d3699c0
	[C]: in function 'assert'
	...an/torch/install/share/lua/5.1/nn/BatchNormalization.lua:77: in function 'checkInputDim'
	...an/torch/install/share/lua/5.1/nn/BatchNormalization.lua:102: in function 'forward'
```

The assertion in the "BatchNormalization.lua" file is
```
assert(iDim == self.nDim or
              (**iDim == self.nDim - 1** and **self.train == false**), string.format(
      'only mini-batch supported (%dD tensor), got %dD tensor instead',
      self.nDim, iDim))
```
According to the assertion. I think that in the TRAIN phase, BatchNormalization required 4D Tensor (a mini-batch of images), but in the TEST phase, it should allow 3D tensor (a single image).

But even when I set the train attribute of **BatchNormalization** to **_false_**.  
`model:get(1):get(2).train = false`

```
nn.SpatialBatchNormalization
{
  gradBias : CudaTensor - size: 32
  output : CudaTensor - size: 200x32x32x32
  bias : CudaTensor - size: 32
  gradInput : CudaTensor - size: 200x32x32x32
  **train : false**
  running_var : CudaTensor - size: 32
  momentum : 0.1
  save_mean : CudaTensor - size: 32
  eps : 1e-05
  _type : torch.CudaTensor
  affine : true
  running_mean : CudaTensor - size: 32
  gradWeight : CudaTensor - size: 32
  save_std : CudaTensor - size: 32
  weight : CudaTensor - size: 32
}
```
The assertion still fails (as above).

Actually, I can solve the problem by reshape from 3D tensor to 4D tensor. 
I am just curious what happen here. 

Thank you very much.

**Comment by soumith** [comment]
maybe because you did not update nn / cunn package. I only added this change 2 days ago.

luarocks install nn
luarocks install cutorch
luarocks install cunn

**Comment by soumith** [comment]
https://github.com/torch/nn/commit/03a46b4413084e0c0a647be2d8c5311c900589f1
